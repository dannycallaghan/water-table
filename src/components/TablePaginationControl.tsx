import { memo, useCallback } from 'react';
import { IPaginationOptions } from '../types/types';

const TablePaginationControl = memo(function TablePaginationControl(props: {
  dataLength: number;
  initialDataLength: number;
  options: IPaginationOptions;
  hideDisplaySelect: boolean;
  filteredDataLength: number;
  handlePaginationChange: (page: number, itemsPerPage: number) => void;
}) {
  const { initialDataLength, handlePaginationChange } = props;
  const {
    options: { page, itemsPerPage },
    filteredDataLength,
  } = props;

  const handleDisplay = useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>): void => {
      const maxPages = Math.ceil(
        initialDataLength / Number(event.target.value)
      );
      let paged = page;
      if ((page as number) > maxPages) {
        paged = maxPages;
      }
      handlePaginationChange(paged, Number(event.target.value));
    },
    [handlePaginationChange, initialDataLength, page]
  );

  const handlePaging = (event: React.MouseEvent<HTMLButtonElement>): void => {
    const pageData = event.currentTarget.dataset['page'];
    const currentPage = props.options.page as number;
    let page;
    switch (pageData) {
      case 'prev':
        page = currentPage - 1;
        break;
      case 'next':
        page = currentPage + 1;
        break;
      default:
        page = Number(pageData);
    }
    if (page < 1) page = 1;
    if (page > Math.ceil(initialDataLength / itemsPerPage)) {
      page = Math.ceil(initialDataLength / itemsPerPage);
    }
    handlePaginationChange(page, itemsPerPage);
  };

  const setButtonState = useCallback(
    (index: number) => {
      const maxPages = Math.ceil(initialDataLength / itemsPerPage);
      const diffBefore = page === maxPages ? 2 : 1;
      const diffAfter = page === 1 ? 2 : 1;

      if (index === page) {
        return 'bg-[#063360] text-white';
      }
      if (index < page - diffBefore || index > page + diffAfter) {
        return 'hidden';
      }
      return 'bg-white';
    },
    [page, initialDataLength, itemsPerPage]
  );

  const getItemInfo = useCallback(() => {
    const lastItem = page * itemsPerPage;
    let lastItemShowing = lastItem;
    if (lastItemShowing > filteredDataLength) {
      lastItemShowing = filteredDataLength;
    }
    const firstItem = lastItem - itemsPerPage;
    return `SHOWING ${firstItem + 1} - ${lastItemShowing} of ${filteredDataLength}`;
  }, [page, itemsPerPage, filteredDataLength]);

  return (
    <>
      <div className="flex flex-row items-center gap-4">
        {!props.hideDisplaySelect && (
          <div className="flex max-w-max flex-col">
            <label
              htmlFor="display-items"
              className="text-[0.7rem] text-gray-400"
            >
              ITEMS PER PAGE
            </label>
            <select
              id="display-items"
              onChange={handleDisplay}
              value={itemsPerPage}
              className="h-8 rounded border border-solid border-slate-300 p-1 text-[0.7rem] text-slate-600 focus-visible:border-slate-300"
            >
              <option value={10}>10</option>
              {initialDataLength > 50 && <option value={50}>50</option>}
              {initialDataLength > 100 && <option value={100}>100</option>}
              {initialDataLength > 500 && <option value={500}>500</option>}
              {initialDataLength > 1000 && <option value={1000}>1000</option>}
              {initialDataLength <= 1000 && (
                <option value={initialDataLength}>All</option>
              )}
            </select>
          </div>
        )}
        <div>
          <p className="text-[0.7rem] uppercase text-gray-400">
            PAGE {page} OF {Math.ceil(filteredDataLength / itemsPerPage)}
          </p>
          <div className="flex flex-row items-center gap-1">
            <button
              disabled={page === 1}
              data-page="prev"
              className="inline-block aspect-square w-8 rounded border border-slate-300 bg-white text-xs disabled:text-slate-300"
              onClick={handlePaging}
            >
              &lt;
            </button>
            {Array.from(
              {
                length: Math.ceil(filteredDataLength / itemsPerPage),
              },
              (_, index) => (
                <button
                  data-page={index + 1}
                  className={`${setButtonState(index + 1)} inline-block aspect-square w-8 rounded border border-slate-300 text-xs disabled:text-slate-300`}
                  key={index + 1}
                  onClick={handlePaging}
                >
                  {index + 1}
                </button>
              )
            )}
            <button
              disabled={page === Math.ceil(filteredDataLength / itemsPerPage)}
              data-page="next"
              className="inline-block aspect-square w-8 rounded border border-slate-300 bg-white text-xs disabled:text-slate-300"
              onClick={handlePaging}
            >
              &gt;
            </button>
          </div>
        </div>
      </div>
      <p className="mt-1 text-[0.7rem] text-gray-400">{getItemInfo()}</p>
    </>
  );
});

export default TablePaginationControl;
