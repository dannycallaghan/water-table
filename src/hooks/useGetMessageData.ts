import { filterByDate, compareISODates } from '../lib/utils';
import { IMessages, IMessageDataOptions, IMessage, IGetMessageDataValues } from '../types/types';

export const defaultMessageDataOptions: IMessageDataOptions = {
  searchTerm: '',
  filter: {
    filterBy: '',
    filterValue: '',
  },
  pagination: {
    itemsPerPage: 100,
    page: 1,
  },
  sortBy: {
    type: 'date',
    key: 'transmittedAt',
    desc: true,
  },
};

function useGetMessageData(
  data: IMessages,
  options: IMessageDataOptions): 
  IGetMessageDataValues {
  let messages = data;

  // Search term
  if (options.searchTerm.length) {
    messages = messages.filter(item => item.sensorId.toLocaleLowerCase().includes(options.searchTerm));
  }

  // Filter
  if (options.filter.filterBy.length) {
    if (options.filter.filterBy === 'date') {
      messages = filterByDate(messages, options.filter.filterValue as string[]);
    }
    if (options.filter.filterBy === 'direction') {
      messages = messages.filter(message => message.direction.toLocaleLowerCase() === (options.filter.filterValue as string).toLowerCase());
    }
  }
  
  // Sort
  messages = messages.sort((a: IMessage, b: IMessage) => {
    switch(options.sortBy.type) {
      case 'number':
        if (options.sortBy.desc) {
          // @ts-expect-error Expecting a number of an arithmetic operation
          return b[options.sortBy.key as keyof IMessage] - a[options.sortBy.key as keyof IMessage];
        }
        // @ts-expect-error Expecting a number of an arithmetic operation
        return a[options.sortBy.key as keyof IMessage] - b[options.sortBy.key as keyof IMessage];
      case 'string':
        if (options.sortBy.desc) {
          // @ts-expect-error Won't allow the use of localeCompare
          return b[options.sortBy.key as keyof IMessage].localeCompare(a[options.sortBy.key as keyof IMessage]);
        }
          // @ts-expect-error Won't allow the use of localeCompare
        return a[options.sortBy.key as keyof IMessage].localeCompare(b[options.sortBy.key as keyof IMessage]);
      default: // date
        if (options.sortBy.desc) {
          return compareISODates(b.transmittedAt.iso, a.transmittedAt.iso);
        }
        return compareISODates(a.transmittedAt.iso, b.transmittedAt.iso);
    }
  });

  // Pagination
  const lastItem = options.pagination.page * options.pagination.itemsPerPage;
  const firstItem = lastItem - options.pagination.itemsPerPage;
  const sliced =  messages.slice(firstItem, lastItem);

  return {
    initialMessagesLength: data.length,
    filteredMessagesLength: messages.length,
    messages: sliced,
  };
}

export default useGetMessageData;
