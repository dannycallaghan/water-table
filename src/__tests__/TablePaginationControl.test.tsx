import { expect, test, describe } from 'vitest';
import { render } from '@testing-library/react';
import TablePaginationControl from '../components/TablePaginationControl.tsx';
import { defaultMessageDataOptions } from '../hooks/useGetMessageData.ts';

describe('TablePaginationControl', () => {
  test('renders correctly', () => {
    const handlePaginationChange = () => {};
    const { asFragment } = render(
      <TablePaginationControl
        initialDataLength={100}
        dataLength={20}
        hideDisplaySelect
        options={{ ...defaultMessageDataOptions.pagination }}
        filteredDataLength={10}
        handlePaginationChange={handlePaginationChange}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
