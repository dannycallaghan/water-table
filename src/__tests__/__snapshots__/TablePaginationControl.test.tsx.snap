// Vitest Snapshot v1

exports[`TablePaginationControl > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row items-center gap-4"
  >
    <div>
      <p
        class="text-[0.7rem] uppercase text-gray-400"
      >
        PAGE 1 OF 1
      </p>
      <div
        class="flex flex-row items-center gap-1"
      >
        <button
          class="inline-block aspect-square w-8 rounded border border-slate-300 bg-white text-xs disabled:text-slate-300"
          data-page="prev"
          disabled=""
        >
          &lt;
        </button>
        <button
          class="bg-[#063360] text-white inline-block aspect-square w-8 rounded border border-slate-300 text-xs disabled:text-slate-300"
          data-page="1"
        >
          1
        </button>
        <button
          class="inline-block aspect-square w-8 rounded border border-slate-300 bg-white text-xs disabled:text-slate-300"
          data-page="next"
          disabled=""
        >
          &gt;
        </button>
      </div>
    </div>
  </div>
  <p
    class="mt-1 text-[0.7rem] text-gray-400"
  >
    SHOWING 1 - 10 of 10
  </p>
</DocumentFragment>
`;
